pwiz_tools/Skyline/TestUtil/ExtensionTestContext.cs

```diff
@@ -59,7 +60,10 @@ public static string GetTestDir(this TestContext testContext)
         {
             // when run with VSTest/MSTest (when .runsettings file is used), use the CustomTestResultsDirectory property if available
             // because there's no other way to override the TestDir
-            return testContext.Properties["CustomTestResultsDirectory"]?.ToString() ?? testContext.TestDir;
+            var path = testContext.Properties["CustomTestResultsDirectory"]?.ToString() ?? testContext.TestDir;

```

Comment:
This seems excessive, since the Docker Containers don't run in the local TestResults folder. It also seems like it would have the effect of multiplying the amount of disk space consumed by the tests running entirely locally without parallel testing, e.g. during nightly builds.

Reply:
Should each test cleanup after itself now (enforced if teamcity test decoration is on)?

Reply:
Don't they already? I don't find any cruft in my TestResults folders after a run.

Reply:
Sorry I meant "shouldn't" (i.e. I thought they already were).

Reply:
I am talking about Nightly and dev runs. Here is WinDirStat on a normal nightly test run:
![image](https://user-images.githubusercontent.com/11541328/209859006-a08f2f53-fe90-4f7a-bdff-9a625dd28c7f.png)
And on a Perf test run:
![image](https://user-images.githubusercontent.com/11541328/209859123-7121628b-5caa-4fcb-a25c-a65afa683e77.png)
Quite a lot of the space in Skyline/bin is consumed by tools folders which are already made language-specific and so multiply by 4:
![image](https://user-images.githubusercontent.com/11541328/209859360-8f67ab71-6ae3-4ea6-82db-df3892ce1bd0.png)
It is my sense that this change will similarly inflate the disk space consumed by nightly runs.

AND, I don't yet feel that the assumptions motivating it are borne out by the way the system works. I want to see better justification and clearer understanding.

Reply:
What's in those TestResult folders? Are they not getting cleaned up because teamcity test decoration is off? (I thought that only changed enforcement, not whether cleanup was actually done, but I haven't looked too closely.)

Brian I assume you were getting reliable failures with multi-language parallel tests before this change, and those failures stopped after the change? I don't understand WHY that would be because I agree with Brendan that those TestResults per language should be isolated in each docker container, but I could be missing some aspect.

Reply:
Yes, I was, lots of cases of (apparently) one test cleaning up the directory while it's in use by another instance of the test - possibly I'm misconfigured somehow? The antivirus thing alone strongly suggests that we've only been getting lucky on timing with parallel tests.

But I'll try to convince myself that there's some other issue here.

Reply:
There is definitely more to this than "getting lucky". I have run these a lot. I guess it does sound like there is something different about how you are running them. Let's start with finding a way for you to reproduce the stability level I currently see on BRENDANX-UW6, our release build computer.

Reply:
And, conversely, let's see if you can go unstable on that machine - all tests on, all languages, running onscreen, with a client count of 6.

Reply:
The antivirus thing probably wasn't luck because inside the docker containers there's no antivirus. That actually is evidence that the docker containers are keeping the non-persistent test results local to the container like I expect them to.

Reply:
Having trouble reproducing the antivirus thing at the moment (really do think its a luck/timing thing) but my latest run has this:
[SkylineTester.log](https://github.com/ProteoWizard/pwiz/files/10316073/SkylineTester.log)


 [11:26] 0.16    SafeDeleteTest                                 (en)   1 failures, 0.94/14.71/52.2 MB, 2/365 handles, 6 sec.

!!! SafeDeleteTest FAILED
Access to the path 'C:\Dev\master_clean\pwiz_tools\Skyline\bin\x64\Release\TestReadOnlyFile.txt' is denied.
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.UnauthorizedAccessException: Access to the path 'C:\Dev\master_clean\pwiz_tools\Skyline\bin\x64\Release\TestReadOnlyFile.txt' is denied.
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.StreamWriter.CreateFile(String path, Boolean append, Boolean checkHost)
   at System.IO.StreamWriter..ctor(String path, Boolean append, Encoding encoding, Int32 bufferSize, Boolean checkHost)
   at System.IO.File.InternalWriteAllText(String path, String contents, Encoding encoding, Boolean checkHost)
   at pwiz.SkylineTest.UtilTest.SafeDeleteTest() in C:\Dev\master_clean\pwiz_tools\Skyline\Test\UtilTest.cs:line 201
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, Binding
Flags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at TestRunnerLib.RunTests.Run(TestInfo test, Int32 pass, Int32 testNumber, String dmpDir, Boolean heapOutput) in C:\Dev\master_clean\pwiz_tools\Skyline\TestRunnerLib\RunTests.cs:line 348
!!!

Reply:
This is an error in Skyline\bin which is a very different location. As I mentioned we already use language separation for tools files in the bin directory. Though, this might also be a test coding error. Why exactly does this need to happen in the bin directory? Or is it just accidentally ending up in the directory of the running application?

Reply:
OK I tested your GetTestDir change and I still get the same ConsoleMultiReplicateImportTest error. Actually it seems to happen more frequently (no idea why).